sender.py

```diff
@@ -0,0 +1,208 @@
+
+
+def main():
+    setup_logging()
+
+    retry_interval = int(os.environ.get('RETRY_INTERVAL', 600))
+    shard, shards = get_sharding_configuration()
+
+    mq_manager = mailqueue.Manager(shard=shard, shards=shards)
+    mq_manager.start()
+
+    for _ in range(DELIVERY_THREADS):
+        d = DeliveryThread(mq_manager, retry_interval)
+        d.start()
+
+    while True:
+        time.sleep(1)
+
+
+class DeliveryThread(threading.Thread):
+    def __init__(self, mq_manager, retry_interval):
+        super().__init__()
+        self.daemon = True
+        self._delivery_agent = DeliveryAgent(mq_manager, get_dns_resolver(), SMTPClient())
+        self._delivery_agent.retry_interval = retry_interval
+
+    def run(self):
+        while True:
+            self._delivery_agent.deliver_single_envelope()
+
+
+class DeliveryAgent:
+    IDLE = 'IDLE (there was no items in the incoming queue)'
+    DONE = 'DONE (there was at least one item to process)'
+
+    def __init__(self, mailqueue, dns_resolver, smtp_client):
+        self.max_delivery_attempts = 4 # 1 initial, 3 retries
+        self.retry_interval = 600 # seconds
+
+        self._dns_resolver = dns_resolver
+        self._mailqueue = mailqueue
+        self._smtp_client = smtp_client
+
+    def deliver_single_envelope(self):
+        envelope = self._mailqueue.get()
+
+        if not envelope:
+            return self.IDLE
+
+        logging.info('Took envelope {} for delivery (domain={}, recipients={})'.format(
+            envelope.id, envelope.destination_domain, envelope.recipients))
+
+        mxs = self._get_mxs(envelope)
+        if not mxs:
+            return self.DONE
+
+        delivered_to = self._try_delivering_to_all_mxs(envelope, mxs)
+        self._handle_attempt_outcome(delivered_to, envelope)
+
+        return self.DONE
+
+    def _get_mxs(self, envelope):
+        try:
+            mxs = self._dns_resolver.get_mxs(envelope.destination_domain)
+            logging.debug('Envelope {}: MXs for {} are {}'.format(envelope.id,
+                                                                  envelope.destination_domain,
+                                                                  mxs))
+            return mxs
+        except TemporaryFailure as e:
+            logging.warning(
+                "Envelope {}: can't resolve MXs for {}: {}".format(envelope.id,
+                                                                   envelope.destination_domain,
+                                                                   e))
+            self._handle_temporary_failure(envelope)
+
+    def _handle_attempt_outcome(self, delivered_to, envelope):
+        if delivered_to:
+            logging.info('Envelope {}: successfully delivered to {}'.format(envelope.id,
+                                                                            delivered_to))
+            self._mailqueue.mark_as_sent(envelope)
+        else:
+            self._handle_temporary_failure(envelope)
```

Comment:
What if some recipients from the envelope were rejected by the remote SMTP server, but others were accepted?